### PR TITLE
swev-id: sphinx-doc__sphinx-9658 restify mocked classes

### LINKED ISSUE #92: Inherited classes not correctly documented when mocked
Inherited classes not correctly documented when mocked

### Describe the bug

We're experiencing an issue when documenting classes that inherit mocked classes. However, classes which inherit other classes from our own package are ok.

This issue appears to be dependent on the `sphinx` version:

- `sphinx<3.0`: Everything is OK. 
- `sphinx>=3.0 < 3.4.2`: Classes that inherit mocked classes are not documented. (see [sphinx #8164](https://github.com/sphinx-doc/sphinx/issues/8164)). This is fixed in `sphinx 3.4.2`. 
- `sphinx>=3.4.2`: The previously missing classes are now documented, but there is a problem with the "Bases" section in the docs. 
 
Example: In the docs for `alibi_detect.utils.pytorch.kernels.DeepKernel` in this readthedocs build https://seldon--338.org.readthedocs.build/projects/alibi-detect/en/338/api/alibi_detect.utils.pytorch.kernels.html, the base class is listed as "Bases: `torch.nn.`" instead of "Bases: `torch.nn.Module`". 

### How to Reproduce

```
$ git clone https://github.com/ascillitoe/alibi-detect.git
$ cd alibi-detect
$ pip install -r requirements/docs.txt
$ make build_docs
$ # open doc/_build/html/api/alibi_detect.utils.pytorch.kernels.html and see "Bases" section.
```

### Expected behavior

The "Bases" section should report `torch.nn.Module` not `torch.nn.`. 

i.e. see
https://seldon--325.org.readthedocs.build/projects/alibi-detect/en/325/api/alibi_detect.utils.pytorch.kernels.html

### Your project

https://github.com/ascillitoe/alibi-detect/tree/feature_sphinx4

### Screenshot with `sphinx==4.2`
sphinx_problem

### Screenshot with `sphinx<3.0`
sphinx_working

### OS

Ubuntu 18.04 (used by readthedocs/build:6.0)

### Python version

3.8.11

### Sphinx version

`>=3.4.2`

### Sphinx extensions

    ["sphinx.ext.autodoc",
    "sphinx.ext.doctest",
    "sphinx.ext.intersphinx",
    "sphinx.ext.todo",
    "sphinx.ext.coverage",
    "sphinx.ext.mathjax",
    "sphinx.ext.ifconfig",
    "sphinx.ext.viewcode",
    "sphinx.ext.napoleon",
    "sphinx_autodoc_typehints",
    "sphinxcontrib.apidoc", 
    "nbsphinx",
    "nbsphinx_link",  
    "myst_parser"]

### Extra tools

_No response_

### Additional context

demo PR:
https://github.com/SeldonIO/alibi-detect/pull/338

readthedocs demo build:
https://seldon--338.org.readthedocs.build/projects/alibi-detect/en/338/api/alibi_detect.utils.pytorch.kernels.html

### Observed failure and reproduction steps
- Observed: "Bases: :py:class:`torch.nn.`" is rendered for classes inheriting a mocked torch.nn.Module, instead of the expected "Bases: :py:class:`torch.nn.Module`".
- Reproduction: Follow the steps above; the incorrect "Bases" line appears in the built HTML.
- Stack trace: None (this is a rendering bug, no crash); the issue manifests as incorrect reference text in the "Bases" section.

### PR DESCRIPTION
**Pull request by casey-brooks**
## Summary
- normalize class name resolution in restify when __qualname__ is empty
- align py36 helper fallback with revised py37 handling
- add unit/functional coverage for mocked bases to prevent regressions

## Testing
- pytest -q

Closes #92

### CODE REVIEW

**Comment by casey-brooks** [comment]
## Test & Lint Summary
- pytest -q
  - 1703 passed, 16 failed, 22 skipped (expected baseline failures: linkcheck counts, autodoc enum/typehint expectations, image converter, setup command options, typing/stringify formatting)
- flake8 sphinx/util/typing.py tests/test_ext_autodoc.py tests/test_util_typing.py tests/roots/test-autodoc-mocked-bases
  - no lint issues

File: sphinx/util/typing.py [hunk 1]
```diff
@@ -208,19 +238,57 @@ def _restify_py37(cls: Optional[Type]) -> str:
...
+            elif not qualname:
+                qualname = repr(cls).replace('typing.', '')
     elif hasattr(cls, '__qualname__'):
-        qualname = '%s.%s' % (module, cls.__qualname__)
+        qual = _normalized(getattr(cls, '__qualname__', None))
+        if not qual:
+            qual = _normalized(getattr(cls, '__name__', None))
+
+        if qual:
+            qualname = '%s.%s' % (module, qual) if module else qual
```
Comment (noa-lucent) [inline sphinx/util/typing.py:274]: [major] The normalization we do here never reaches the renderer on Python <3.7. A few lines below we still have the legacy `elif hasattr(cls, '__qualname__')` branch returning `cls.__qualname__` verbatim, so mocked classes with an empty qualname still produce `:py:class:`torch.nn.` when `_restify_py36` runs. Please rework `_restify_py36` so the final branch reuses the normalized name (mirroring the py37 helper) instead of the raw attribute; otherwise the issue persists on the py36 code path.

**Review by noa-lucent — CHANGES_REQUESTED** [review]
Blocking on `_restify_py36`: the new normalization never reaches the rendering branch, so mocked bases still come out as `:py:class:`torch.nn.` on the py36 helper. Please route that fallback through the normalized name (and add coverage) so the fix applies to both helpers.

**Comment by casey-brooks** [comment]
## Test & Lint Summary
- pytest tests/test_util_typing.py::test_restify_py36_mocked_class_falls_back_to_name -q
  - 1 passed, 0 failed, 0 skipped (py36 fallback coverage)
- flake8 sphinx/util/typing.py tests/test_util_typing.py
  - no lint issues

**Review by noa-lucent — APPROVED** [review]
Looks good—thanks for rerouting the py36 helper through the normalized name and covering both restify paths.
